### PR TITLE
`Development`: Remove redundant latest-due-date calls

### PR DESCRIPTION
**Pull request by simon-hng**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

### Motivation and Context

Reduces work on the server and enables future caching client side.

### Description

- We cache the calls to `latest-due-date` in a service that is provided in the parent component, leading to reduced rest calls
- We add component scoped providers of those cache services where we want caching

### Testing steps
Prerequisites:
- Exercise where now is after due date
- 2 Submissions for this exercise

Steps:
- Have a look at result for submission1. There should be no text about testcase details.
- Set individual due date of submission2 to a date in the future (e.g. in an incognito browser window)
- Look at the result for submission1 again after in and out of the page (or reloading)

#### Code Review
- [x] Review 1
- [x] Review 2

#### Screenshots
Text mentioned above: `Even after the due date not for all test cases a detailed feedback is shown. This could be the case if other students are still working on the exercise (e.g. to allow for the compensation of sick days)...`
Screenshot_20230131_145014

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/result/result.component.ts [hunk 1]
```diff
@@ -266,16 +268,15 @@ export class ResultComponent implements OnInit, OnChanges {
      */
     private determineShowMissingAutomaticFeedbackInformation(componentInstance: FeedbackComponent) {
         if (!this.latestIndividualDueDate) {
-            this.exerciseService.getLatestDueDate(this.exercise!.id!).subscribe((latestIndividualDueDate?: dayjs.Dayjs) => {
-                this.latestIndividualDueDate = latestIndividualDueDate;
-                this.initializeMissingAutomaticFeedbackAndLatestIndividualDueDate(componentInstance);
-            });
-        } else {
-            this.initializeMissingAutomaticFeedbackAndLatestIndividualDueDate(componentInstance);
+            if (this.exerciseCacheService) {
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/result/result.component.ts:271]: only use caching if a caching service is provided in a parents provider array.

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts [hunk 2]
```diff
@@ -42,6 +43,7 @@ enum FilterProp {
     selector: 'jhi-exercise-scores',
     styleUrls: ['./exercise-scores.component.scss'],
     templateUrl: './exercise-scores.component.html',
+    providers: [ExerciseCacheService],
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts:46]: We enable caching on a component basis by adding the cache service to the providers

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts [hunk 3]
```diff
@@ -64,7 +64,7 @@ export class FeedbackComponent implements OnInit {
      * the result.
      */
     @Input() showMissingAutomaticFeedbackInformation = false;
-    @Input() latestIndividualDueDate?: dayjs.Dayjs;
+    @Input() latestDueDate?: dayjs.Dayjs;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:67]: Why did you rename this? I think it's more clearer with the 'individual'

File: src/main/webapp/app/exercises/shared/result/result.component.ts [hunk 4]
```diff
@@ -78,6 +79,7 @@ export class ResultComponent implements OnInit, OnChanges {
         private http: HttpClient,
         private modalService: NgbModal,
         private exerciseService: ExerciseService,
+        @Optional() private exerciseCacheService: ExerciseCacheService,
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/result/result.component.ts:82]: Why is this optional?

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts (hunk 3 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:67]: If you look at the rest call serverside you see that it is either the lastest individual due date or the latest exercise due date

File: src/main/webapp/app/exercises/shared/result/result.component.ts (hunk 4 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/result/result.component.ts:82]: We do not always want to cache. The cache service is provided on a component basis and not in the root. i.e. if you navigate out of a cached component the cache is not persistent

**Review by TheZoker — APPROVED** [review]
Code 👍

**Review by JohannesStoehr — APPROVED** [review]
Thanks for answering my questions. Code changes look good to me

**Comment by simon-hng** [comment]
@krusche I will take this PR to a testing session next week

**Review by luisa1999 — APPROVED** [review]

**Review by nikirovast — APPROVED** [review]
Tested it during test-session, seems to work.

**Review by frederik-kt — APPROVED** [review]
Tested on TS1. Message is shown as expected.

**Review by manuelmanso — APPROVED** [review]
Code LGTM. Tested in the testing session and worked.

**Review by danielbode — APPROVED** [review]
Tested on ts 1. Works as expected.

**Review by Vinicius-Agreste — APPROVED** [review]
Tested on ts1.
